File: gobang/gobang_c++/AI/static.cpp
//
// Created by ycm on 2023/12/8.
//

#include "static.h"

std::string Leaf::toString() const {
    std::string result = "Leaf(score=";
    result += std::to_string(this->score);
    result += ", step=";
    result += std::to_string(this->step);
    result += ")";

    return result;
}

std::string Place::toString() const {
    std::string result = "Place(";
    result += std::to_string(this->x);
    result += ", ";
    result += std::to_string(this->y);
    result += ", role=";
    switch (this->role) {
        case Role::HUMAN:
            result += "human";
            break;
        case Role::COMPUTER:
            result += "computer";
            break;
        case Role::NONE:
            result += "blank";
            break;
        default:
            result += "unknown";
            break;
    }
    if(this->score!=0)result += ", score="+std::to_string((long)this->score);
    if(this->scoreCom!=0)result += ", computer's score="+std::to_string((long)this->scoreCom);
    if(this->scoreHum!=0)result += ", human's score="+std::to_string((long)this->scoreHum);
    if(this->step)result += ", step="+std::to_string(this->step);
    if(this->min)result += ", min="+std::to_string(this->min);
    result += ")";

    return result;
}

File: gobang/gobang_python/AI/static.py
import enum


class Color(enum.IntEnum):
    BLANK = 2
    BLACK = 0
    WHITE = 1


class Char(enum.Enum):
    BLANK = " "
    BLACK = "■"
    WHITE = "o"


class Role(enum.IntEnum):
    HUMAN = 0
    COMPUTER = 1
    BLANK = 2

    def name(self) -> str:
        return ("HUMAN", "COMPUTER", "BLANK")[self]


class Leaf:
    def __init__(self, score, step=None, steps=None):
        self.score = score
        self.step = step
        # self.steps = steps
        self.abCut = False
        self.depth = 0

    def __repr__(self):
        return f"Leaf({self.score}, {self.step})"  # , {self.steps})"


class Place:
    def __init__(self, x, y, v: Leaf = None, role: Role = Role.BLANK):
        self.x = x
        self.y = y
        self.v = v
        self.role = role
        self.score = self.scoreCom = self.scoreHum = 0
        self.step = self.min = 0

    def __str__(self):
        result = f"Place({self.x}, {self.y}, role={self.role.name()}"
        if self.score:
            result += f", score={self.score}"
        if self.scoreCom:
            result += f", Computer's score={self.scoreCom}"
        if self.scoreHum:
            result += f", Human's score={self.scoreHum}"
        if self.step:
            result += f", step={self.step}"
        if self.min:
            result += f", min={self.min}"
        result += ")"
        return result

    def __repr__(self):
        result = f"Place({self.x}, {self.y}"
        if self.role != Role.BLANK:
            result += f", role={self.role.name()}"
        result += ")"
        return result

    def __iter__(self):
        yield self.x
        yield self.y

    def __getitem__(self, key):
        if key == 0:
            return self.x
        elif key == 1:
            return self.y
        else:
            raise KeyError(f"bad key '{key}'")

    def __hash__(self):
        return hash((self.x, self.y))

    def __eq__(self, other):
        try:
            it = iter(other)
            x = next(it)
            y = next(it)
            return self.x == x and self.y == y
        except Exception:
            return NotImplemented


class Score(enum.IntEnum):
    ONE = 10
    TWO = 100
    THREE = 1000
    FOUR = 10000
    FIVE = 100000
    BLOCKED_ONE = 1
    BLOCKED_TWO = 10
    BLOCKED_THREE = 100
    BLOCKED_FOUR = 10000
